src/pg_query_summary_truncate.c

```diff
@@ -0,0 +1,409 @@
+#include "pg_query.h"
+#include "pg_query_internal.h"
+#include "pg_query_summary.h"
+
+#include "nodes/pg_list.h"
+#include "nodes/nodeFuncs.h"
+
+enum TruncationAttr {
+	TRUNCATION_TARGET_LIST,
+	TRUNCATION_WHERE_CLAUSE,
+	TRUNCATION_VALUES_LISTS,
+	TRUNCATION_COLS,
+	TRUNCATION_CTE_QUERY,
+};
+
+typedef struct {
+	List *truncations;
+	int32_t depth;
+} TruncationState;
+
+typedef struct {
+	enum TruncationAttr attr;
+	Node *node;
+	int32_t depth;
+	int32_t length;
+} PossibleTruncation;
+
+static bool summary_truncation_options(Node *node, TruncationState *state);
+static void sort_truncations(Node *node, TruncationState *state);
+static void apply_truncations(Node *node, TruncationState *state);
+
+static int32_t select_target_list_len(List *nodes);
+static int32_t select_values_lists_len(List *nodes);
+static int32_t update_target_list_len(List *nodes);
+static int32_t where_clause_len(Node *node);
+static int32_t cols_len(List *nodes);
+
+/*
+ * Given a walked parse tree and summary, store the truncated version in `summary`.
+ *
+ * Returns NULL on success.
+ */
+PgQueryError *pg_query_summary_truncate(Summary *summary, Node *node)
+{
+	TruncationState state = {NULL, 0};
+
+	printf("\n\n! pg_query_summary_truncate()\n");
+	summary_truncation_options(node, &state);
+	return NULL;
+}
+
+static char *truncate_str(char *str, size_t max_chars)
+{
+	// FIXME: This WILL blow up for multi-byte UTF-8 characters, since it's going
+	// by byte instead of character. I think C has functions to deal with this,
+	// I just haven't learned about them yet. -duckinator
```

Comment:
It looks like we can use `pg_mbcliplen` for this, which lets you specify the number of bytes you want to limit to, but respects multi-byte boundaries: https://github.com/postgres/postgres/blob/master/src/backend/utils/mb/mbutils.c#L1084

Alternatively, I wonder if we should treat the truncation limit as number of multi-byte characters, not number of bytes. In that case, `pg_mbcharcliplen` could be used: https://github.com/postgres/postgres/blob/master/src/backend/utils/mb/mbutils.c#L1126 (and we should make sure to use `pg_mbstrlen` to get calculate the length of strings)